src/server/gc_worker/compaction_filter.rs

```diff
@@ -100,6 +105,29 @@ impl CompactionFilterFactory for WriteCompactionFilterFactory {
             &*gc_context.cfg_tracker.value(),
             &gc_context.cluster_version,
         ) {
+            debug!("skip gc in compaction filter because it's not allowed");
+            return std::ptr::null_mut();
+        }
+
+        let ratio_threshold = gc_context.cfg_tracker.value().ratio_threshold;
+        let (mut needs_gc, mut mvcc_props) = (false, MvccProperties::new());
+        for i in 0..context.file_numbers().len() {
+            let table_props = context.table_properties(i);
+            let user_props = unsafe {
+                &*(table_props.user_collected_properties() as *const _
+                    as *const RocksUserCollectedPropertiesNoRc)
+            };
+            if let Ok(props) = RocksMvccProperties::decode(user_props) {
+                mvcc_props.add(&props);
+                let sst_ratio = ratio_threshold + SINGLE_SST_RATIO_THRESHOLD_ADJUST;
+                if check_need_gc(safe_point.into(), sst_ratio, &props) {
+                    needs_gc = true;
+                    break;
+                }
+            }
+        }
+        if !needs_gc && !check_need_gc(safe_point.into(), ratio_threshold, &mvcc_props) {
```

Comment:
even when `check_need_gc==false`, if it is bottom level compaction, we can still run compaction filter?

Reply:
Have commentted for this.